lib/deploy/functions/release.js

```diff
@@ -110,123 +117,53 @@ module.exports = function(context, options, payload) {
+    };
+    var printFail = function(op) {
+      _endTimer(op.functionName);
+      failedDeployments += 1;
+      utils.logWarning(chalk.bold.yellow('functions[' + op.functionName + ']: ') + 'Deployment error.');
+      if (op.error.code === 8) {
+        logger.debug(op.error.message);
+        logger.info('You have exceeded your deployment quota, please deploy your functions in batches by using the --only flag, ' +
+          'and wait a few minutes before deploying again. Go to https://firebase.google.com/docs/cli/#partial_deploys to learn more.');
       } else {
-        utils.logSuccess(chalk.bold.green('functions: ') + successfulOps.length + ' function(s) deployed successfully.');
+        logger.info(op.error.message);
       }
-    }
-    return failed ? RSVP.reject() : RSVP.resolve();
-  }
-
-  function _nameFromOperation(operation) {
-    if (!operation || !_.isString(operation.func)) {
-      return null;
-    }
-
-    return operation.func.split('/')[5];
-  }
-
-  function _pollOperations(successfulCalls) {
-    function logOps(ops) {
-      _.forEach(ops, function(operation) {
-        var functionName = _nameFromOperation(operation);
-        var msg = '[functions] operation poll: '
-          + functionName + ': '
-          + operation.type + ' ' + operation.name + ' is '
-          + (operation.done ? 'done.' : 'not done.');
-        logger.debug(msg);
-      });
-    }
-    return new RSVP.Promise(function(resolve, reject) {
-      function poll() {
-        var unfinishedOps = _.chain(successfulCalls).filter(function(operation) {
-          return !(operation.done || operation.error);
-        }).map(function(operation) {
-          return gcp.cloudfunctions.check(operation).then(function(op) {
-            var name = _nameFromOperation(op);
-            if (name && op.done) {
-              _endTimer(name);
-            }
-            return op;
-          });
-        }).value();
-
-        if (unfinishedOps.length === 0) {
-          return resolve();
-        }
-
-        RSVP.allSettled(unfinishedOps).then(function(checks) {
-          var successfulChecks = _.chain(checks).filter({'state': 'fulfilled'}).map('value').value();
-          logOps(successfulChecks);
-
-          var allDone = _.every(successfulChecks, { done: true });
-          if (allDone) {
-            return resolve();
-          }
-          setTimeout(poll, POLL_INTERVAL);
-        }).catch(function(err) {
-          return reject(err);
-        });
+    };
+    var retryCondition = function(result) {
+      // The error codes from a Google.LongRunning operation follow google.rpc.Code format.
+      var doNotRetryCodes = [
+        3,  // invalid argument
+        5,  // not found
+        6,  // already exists
+        7,  // permission denied
+        8,  // resource exhausted (quota issue)
+        9,  // failed precondition
+        11, // out of range
+        12, // unimplemented
+        16  // unauthenticated
+      ];
```

Comment:
It won't loop infinitely since failed deployments only gets tried once. But yes I agree a whitelist is better than a blacklist, I will make the change. The error that gets printed out comes directly from the HTTPS response, which does have a readable message.